### PR TITLE
Drop Pydantic v1

### LINKED ISSUE #91: Remove pydantic v1 support
Does anyone need Pydantic < 1 support? @fjclark or @jmorado ? It'd be great to drop this in the tests.

### PR DESCRIPTION
**Pull request by mattwthompson**
## Description

Fixes #91 

## Status
- [x] Ready to go

### CODE REVIEW

**Review by j-wags — CHANGES_REQUESTED** [review]
I didn't find any problems myself, but I let claude do a once-over and it spotted a sneaky regression wrt vsites and cosmetic attributes. I worked with claude to generate minimal tests expressing this behavior.

I lean slightly toward favoring fixing this regression (since it could cause user surprise) but I'd also be fine with A) just listing it in the release notes (and removing the new tests I committed) or B) erroring in cases where it could make a difference (though this could be hard to detect)

File: descent/tests/test_train.py [hunk 1]
```diff
@@ -474,3 +474,67 @@ def test_init_vsites_regularization(
...
+    # key a user builds from a SMIRKS -- which sets neither -- must still match
+    # the force field's key in ``include`` / ``exclude``.
+
+    def test_init_vsites_exclude_key_without_virtual_site_type(self, water_sites_ff):
+        # A user references the lone pair the way its key prints, i.e. by
+        # SMIRKS + name + match, without the interchange-only virtual_site_type.
+        excluded = openff.interchange.models.PotentialKey(
+            id="[#1:2]-[#8X2H2+0:1]-[#1:3] EP once",
+            associated_handler="VirtualSites",
+        )
```
Comment (mattwthompson) [inline descent/tests/test_train.py:489]: This is now how as-applied virtual site parameters are stored in Interchange. There is a separate class for virtual sites that intentionally stores relevant fields:

```
In [1]: from openff.toolkit import Molecule, ForceField

In [2]: virtual_site_keys = [*ForceField("opc.offxml").create_interchange(Molecule.from_smiles("O").
      ⋮ to_topology())['VirtualSites'].key_map]

In [3]: virtual_site_keys
Out[3]: [SMIRNOFFVirtualSiteKey with orientation atom indices (0, 1, 2)]

In [4]: type(virtual_site_keys)
Out[4]: list

In [5]: [type(key) for key in virtual_site_keys]
Out[5]: [openff.interchange.models.SMIRNOFFVirtualSiteKey]

In [6]: virtual_site_keys[-1].dict()
<ipython-input-6-9d453a18a1da>:1: PydanticDeprecatedSince20: The `dict` method is deprecated; use `model_dump` instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.13/migration/
  virtual_site_keys[-1].dict()
Out[6]:
{'atom_indices': None,
 'orientation_atom_indices': (0, 1, 2),
 'type': 'DivalentLonePair',
 'name': 'EP',
 'match': 'once'}
```

This approach, be it by Claude or our scientists, of smooshing together 3 parameters into a single string is not encouraged

File: descent/train.py [hunk 2]
```diff
@@ -418,7 +349,10 @@ def _prepare_vsites(
         # avoids hard-coding them.
         vsite_cols = list(force_field.v_sites.default_units().keys())
 
-        all_keys = [_PotentialKey(**key.dict()) for key in force_field.v_sites.keys]
+        all_keys = [
+            openff.interchange.models.PotentialKey(**key.dict())
+            for key in force_field.v_sites.keys
+        ]
```
Comment (mattwthompson) [inline descent/train.py:390]: Ideally switch these over to `VirtualSiteKey`s

File: descent/train.py
Comment (mattwthompson) [inline descent/train.py]: Note to self: this is the point of failure. It's not actually related to Pydantic API changes, it's related to Interchange behavior changes over some period of years. The ~vendored code here has different behavior than upstream.

If we want exclude a fairly generic parameter

```python
ipdb> [e.dict() for e in excluded_keys]
[{'id': '[#6X4:1]-[#6X4:2]', 'mult': None, 'associated_handler': 'Bonds', 'bond_order': None, 'virtual_site_type': None, 'cosmetic_attributes': {}}]
```

but a similar-looking parameter also has cosmetic attributes:
```python
ipdb> [u.dict() for u in unfrozen_keys if u.cosmetic_attributes]
[{'id': '[#6X4:1]-[#6X4:2]', 'mult': None, 'associated_handler': 'Bonds', 'bond_order': None, 'virtual_site_type': None, 'cosmetic_attributes': {'foo': 'bar'}}]
```

then nothing is excluded (probably one parameter should be):
```python
ipdb> len(unfrozen_keys)
2
ipdb> len({key_to_row[key] for key in unfrozen_keys if key not in excluded_keys})
2
```
Optionally add a breakpoint before returning and run something like 

```
$ pixi run -e dev python -m pytest -v descent/tests/test_train.py::TestTrainable::test_init_exclude_key_without_cosmetic_attributes --pdb -s

File: descent/train.py (hunk 2 above)
Comment (fjclark) [reply descent/train.py:390]: Apologies for wading in here, but I'd like to understand this. My understanding is: we want the keys to match force field parameters independent of which atoms they're applied to, but `VirtualSiteKey` is a subclass of `TopologyKey`, so identifies where parameters are matched rather than the parameters themselves. `PotentialKey`s identify parameters, independent of where they're applied. So how can we replace `PotentialKey`s with `VirtualSiteKey`s? 

Also, from above, is the "smooshing together" of parameters not a fairly old feature of Interchange? https://github.com/openforcefield/openff-interchange/blob/100e7f09923c8f42a64672dade85a5e9aa256a2c/openff/interchange/smirnoff/_virtual_sites.py#L121

Thanks!

Comment (mattwthompson) [reply descent/train.py:390]: You're right - I confused myself with these classes and what they do. I'll take a stab at fixing this regression - would you be up for reviewing a fix for this? ETA mid next week

I do think there could be a better solution than stringifying the relevant fields together, but that's what is done in Interchange and ought to be what this code builds off of.

Comment (fjclark) [reply descent/train.py:390]: Yes, certainly happy to review.

**Comment by j-wags** [comment]
Thanks Finlay - let's consider your review final for this PR. 

Just writing this here for the record - my previous "request changes" review is not blocking merging this PR, I just can't find a way to remove it. So if Finlay approves then this is good to go!

**Comment by mattwthompson** [comment]
Jeff + Claude identified two cases in which the current changes fail. They are sane tests which produce genuine failures stemming from some dusty edge cases deep inside our codebase. After some thought, though, I think they're outside the scope of reasonable user behavior / use cases

1. Using cosmetic attributes in `ParameterConfig.exclude` makes things very funky - most of the time makes them useless, in strange edge cases causes weird collisions. But why would the user actually want to define excluded keys by this means?
2. If, also in `ParameterConfig.exclude`, virtual sites parameters stored without explicitly defining the virtual site type, there are some surprising failures. This is a pretty strange case, though - potential keys for virtual sites without the `.virtual_site_type` defined is easy to call invalid and a bug in Interchange[^1] - and I'm inclined to suggest that this is also an invalid input.

I can easily work around my opinions here and make the exclusion logic more robust to these details, basically just not consider these fields when doing the check[^2]. But I'm inclined to conclude that these tests are reasonable and valid but not quite what we should be focusing on. What do you think @fjclark?

[^1]: Wiring this through to system creation would cause either an error I can't remember writing or a silent non-failure that should be loud. Internally, these potential keys are never created with `.associated_handler="VirtualSites"` and also `.virtual_site_type=None` so arguably this should be caught at validation time.
[^2]: The buck stops with default Pydantic model equality, which just compares all values of all fields and fails if there's any mismatch

File: descent/train.py [hunk 3]
```diff
@@ -50,54 +50,22 @@ def _validate_trainable_keys(
...
-                and self.id == other.id
-                and self.mult == other.mult
-                and self.associated_handler == other.associated_handler
-                and self.bond_order == other.bond_order
-            )
-
-    def _convert_keys(value: typing.Any) -> typing.Any:
-        if not isinstance(value, list):
-            return value
+def _convert_keys(value: typing.Any) -> typing.Any:
```
Comment (fjclark) [inline descent/train.py]: Is this function not redundant now? Previously it would convert interchange PotentialKeys to descent's `_PotentialKey` class, but now it seems to convert interchange PotentialKeys back into themselves. If I understand correctly, there's no case where this function would change its input.

**Review by fjclark — CHANGES_REQUESTED** [review]
In response to @mattwthompson's comment above:

My understanding is that these tests point to a meaningful regression (or at least change) which might trip up currently working fitting pipelines (because currently excluded PotentialKeys could now be included). If I understand correctly, we also agree that cosmetic attributes and `virtual_site_type` shouldn't have to be defined by the user in `ParameterConfig.exclude` to match. So, I lean slightly towards keeping these tests and fixing them by ignoring these fields when checking.

However, I think a better behaviour in general might be to only compare keys which are set in the `ParameterConfig.exclude`, rather than manually specifying a list of keys to check for equality, e.g. a torsion set with id but no multiplicity specified would match all the torsion terms with that id and any multiplicity. That would allow the user to dial up or down the specificity easily, and would have saved me some time in the past where I tried to exclude torsions but forgot I needed to specify `mult`. This is probably out of scope here, but could be nice to see in future.

In general, it would be great if it was easy to specify exclusions and inclusions just from a look at the force field, which is not the case for the "smooshed together" virtual site id, but I'm not sure how best to fix that.

Just requesting changes as I think `_convert_keys` may be redundant.

File: descent/train.py (hunk 3 above)
Comment (mattwthompson) [reply descent/train.py]: Yup, forgot to drop this

**Comment by mattwthompson** [comment]
Great, I'll run the equality check through a filtering function that's smarter than just checking raw equality

Revamping `ParameterConfig.exclude` sound worthwhile but let's iterate on that once I drop all of the Pydantic v1 code here

File: .github/codecov.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+coverage:
+  status:
+    project:
+      default:
+        target: 95%
```
[the remark this answers is not in the document]
Comment (mattwthompson) [reply .github/codecov.yaml:5]: Read the docs https://docs.codecov.com/docs/codecov-yaml#can-i-name-the-file-codecovyml

**Comment by mattwthompson** [comment]
I think this is ready for review again. The important changes I've made today are:

* 1fc145f8878733a305e458aadf0c263d98ea6de6 ignores `PotentialKey.cosmetic_attributes` and `PotentialKey.virtual_site_type`[^1] when checking if an applied key is in the exclusion list[^2]
* ad6751f65363ffc8a99d9ecc111f49398da28df9 removes the dead `_convert_keys` function/validator, so diff coverage should be 100%
* 48441bd8dccfc4a7b31532a35a24cbafee4f4a67 uses `model_dump` instead of `dict`, which is "deprecated"

I also asked Copilot to check it over and it found nothing I thought was correct + in scope here

[^1]: Indifferent on this one, and this type of key construction is liable to fail validation in Interchange in the future
[^2]: The ignorance happens on both the excluded keys and the keys we care about, which the Claude-aided tests helpfully cover both of

File: descent/train.py [hunk 5]
```diff
@@ -241,12 +182,30 @@ def _prepare_rows(
         assert all_keys is not None
 
         excluded_keys = config.exclude or []
+        # TODO: Possibly strip irrelevant fields from included keys as well
```
Comment (fjclark) [inline descent/train.py]: On the Copilot comment:

> _prepare_rows strips virtual_site_type / cosmetic_attributes only for the exclude check, but still uses the full PotentialKey objects as dict keys in key_to_row. If a user supplies include keys that omit those interchange-populated fields (the same scenario covered by the new exclude tests), key_to_row[key] can fail to resolve the row index or behave inconsistently. Normalize keys consistently (e.g., strip before building the mapping and before lookup) so include and exclude use the same matching semantics.

My understanding is that this would be a failure case -- I'll write a test to show this.

**Review by fjclark — CHANGES_REQUESTED** [review]
Thanks! I think the copilot comment highlighted a genuine failure case. I'll write a test to illustrate then we can think about the best solution.

**Comment by fjclark** [comment]
OK, I've opened a PR to `no-pydantic-v1` illustrating the issue and proposing a fix: https://github.com/openforcefield/descent/pull/108

**Comment by mattwthompson** [comment]
I think that solution is good - better than putting it off until later - and I'll just merge it into here

**Comment by fjclark** [comment]
On ignoring the `virtual_site_type`, I'm also conflicted. It seems like it should be treated in the same way as other fields like `mult`, because it is required to specify the physics (unlike cosmetic attributes). But then, it ignoring it seems more consistent with how interchange treats it in a few ways -- it's not included in `__hash__` and `__repr__` (should they be?) -- and keeps it backwards compatible.

My slight preference is keeping it ignore for now for backwards compatibility, then updating the permissive eq function (in another PR) so that any field can be left as `None` (and ignored in the comparison), then we'll have both backwards compatibility and consistent treatment of all the fields.

**Comment by fjclark** [comment]
Also, unrelated side question: what's OpenFF's/ OMSF's policy on AI models adding themselves to commits? Feel free to ignore here as irrelevant.

**Comment by mattwthompson** [comment]
> On ignoring the `virtual_site_type`,

I'm fine with either approach; the user base for this package is (currently) minuscule and exclusively power users so I'm less concerned about breakages and pointy edges than a large-scale product that's exposed to beginner. Later we probably should revisit the object's behavior in Interchange as it doesn't make much sense to define an applied parameter as being associated with a virtual site type but allow that field to be none and/or ignore it in some comparisons. (Alas, this case has probably never been thought of before this week)

I do have a preference for keeping things moving forward, and given that I'm fine with most paths here I'm glad to defer to your preference

> OpenFF's/ OMSF's policy on AI models adding themselves to commits

There is no policy, at least right now

File: descent/train.py [hunk 6]
```diff
@@ -228,25 +144,65 @@ class Trainable:
...
-        unfrozen_keys = config.include or all_keys
+        assert len(all_keys) == len(set(all_keys)), "duplicate keys found"
 
-        key_to_row = {key: row_idx for row_idx, key in enumerate(all_keys)}
-        assert len(key_to_row) == len(all_keys), "duplicate keys found"
+        unfrozen_rows = set()
+        for i, key in enumerate(all_keys):
+            if config.include:
+                if not any(
+                    Trainable._permissive_key_eq(key, included_key)
```
[the remark this answers is not in the document]
Comment (fjclark) [reply descent/train.py]: My thinking was that this won't be possible if we add fancier comparison logic in future, and that the added cost should be relatively small.

[the remark this answers is not in the document]
Comment (mattwthompson) [reply descent/train.py]: Not concerned about this right now

**Comment by fjclark** [comment]
I think the Copilot hidden comment is valid; I'll address.

**Comment by fjclark** [comment]
OK, fix to Copilot comment here: https://github.com/openforcefield/descent/pull/109

**Review by fjclark — APPROVED** [review]
Looks good to me!
